File: src/main/java/me/syari/ss/discord/internal/requests/RestAction.java
package me.syari.ss.discord.internal.requests;

import me.syari.ss.discord.api.exceptions.ErrorResponseException;
import me.syari.ss.discord.api.exceptions.RateLimitedException;
import me.syari.ss.discord.api.requests.Request;
import me.syari.ss.discord.api.requests.Response;
import me.syari.ss.discord.api.requests.RestFuture;
import me.syari.ss.discord.internal.JDA;
import okhttp3.RequestBody;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.function.BiFunction;
import java.util.function.Consumer;

public class RestAction<T> {
    private static final Consumer<Object> DEFAULT_SUCCESS = o -> {
    };

    private static final Consumer<? super Throwable> DEFAULT_FAILURE = t -> {
    };

    protected final JDA api;

    private final Route route;
    private final BiFunction<Response, Request<T>, T> handler;

    public RestAction(@NotNull JDA api, Route route) {
        this(api, route, null);
    }

    public RestAction(JDA api, Route route, BiFunction<Response, Request<T>, T> handler) {
        this.api = api;
        this.route = route;
        this.handler = handler;
    }

    @NotNull
    public JDA getJDA() {
        return api;
    }

    public void queue() {
        Route route = finalizeRoute();
        RequestBody data = finalizeData();
        api.getRequester().request(new Request<>(this, DEFAULT_SUCCESS, DEFAULT_FAILURE, true, data, route));
    }

    @NotNull
    private CompletableFuture<T> submit(boolean shouldQueue) {
        Route route = finalizeRoute();
        RequestBody data = finalizeData();
        return new RestFuture<>(this, shouldQueue, data, route);
    }

    public T complete() {
        try {
            return complete(true);
        } catch (RateLimitedException e) {
            throw new AssertionError(e);
        }
    }

    public T complete(boolean shouldQueue) throws RateLimitedException {
        if (CallbackContext.isCallbackContext()) {
            throw new IllegalStateException("Preventing use of complete() in callback threads! This operation can be a deadlock cause");
        }
        try {
            return submit(shouldQueue).get();
        } catch (Throwable e) {
            if (e instanceof ExecutionException) {
                Throwable t = e.getCause();
                if (t instanceof RateLimitedException) {
                    throw (RateLimitedException) t;
                } else if (t instanceof ErrorResponseException) {
                    throw (ErrorResponseException) t;
                }
            }
            throw new RuntimeException(e);
        }
    }

    protected RequestBody finalizeData() {
        return null;
    }

    protected Route finalizeRoute() {
        return route;
    }

    public void handleResponse(@NotNull Response response, Request<T> request) {
        if (response.isOk()) {
            handleSuccess(response, request);
        } else {
            request.onFailure(response);
        }
    }

    protected void handleSuccess(Response response, Request<T> request) {
        if (handler == null) {
            request.onSuccess(null);
        } else {
            request.onSuccess(handler.apply(response, request));
        }
    }
}

File: src/main/java/me/syari/ss/discord/internal/requests/RestAction.kt
package me.syari.ss.discord.internal.requests

import me.syari.ss.discord.api.exceptions.ErrorResponseException
import me.syari.ss.discord.api.exceptions.RateLimitedException
import me.syari.ss.discord.api.requests.Request
import me.syari.ss.discord.api.requests.Response
import me.syari.ss.discord.api.requests.RestFuture
import me.syari.ss.discord.internal.JDA
import me.syari.ss.discord.internal.requests.CallbackContext.Companion.isCallbackContext
import okhttp3.RequestBody
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.function.BiFunction
import java.util.function.Consumer

open class RestAction<T>(
    val jda: JDA, private val route: Route, private val handler: ((Response, Request<T>) -> T)?
) {
    constructor(api: JDA, route: Route): this(api, route, null)

    fun queue() {
        val data = finalizeData()
        val request = Request(this, DEFAULT_SUCCESS, DEFAULT_FAILURE, true, data, route)
        jda.requester.request(request)
    }

    private fun submit(shouldQueue: Boolean): CompletableFuture<T> {
        val data = finalizeData()
        return RestFuture(this, shouldQueue, data, route)
    }

    fun complete(): T {
        return try {
            complete(true)
        } catch (e: RateLimitedException) {
            throw AssertionError(e)
        }
    }

    @Throws(RateLimitedException::class)
    fun complete(shouldQueue: Boolean): T {
        check(!isCallbackContext) { "Preventing use of complete() in callback threads! This operation can be a deadlock cause" }
        return try {
            submit(shouldQueue).get()
        } catch (e: Throwable) {
            if (e is ExecutionException) {
                val t = e.cause
                if (t is RateLimitedException || t is ErrorResponseException) {
                    throw t
                }
            }
            throw RuntimeException(e)
        }
    }

    open fun finalizeData(): RequestBody? {
        return null
    }

    private fun finalizeRoute(): Route? {
        return route
    }

    open fun handleResponse(
        response: Response, request: Request<T>
    ) {
        if (response.isOk) {
            handleSuccess(response, request)
        } else {
            request.onFailure(response)
        }
    }

    open fun handleSuccess(
        response: Response, request: Request<T>
    ) {
        if (handler == null) {
            request.onSuccess(null)
        } else {
            request.onSuccess(handler.invoke(response, request))
        }
    }

    companion object {
        private val DEFAULT_SUCCESS = Consumer { _: Any? -> }
        private val DEFAULT_FAILURE: Consumer<in Throwable> = Consumer { }
    }
}